lib/hotel.rb

```diff
@@ -0,0 +1,102 @@
+module Hotel
+  class Hotel
+    attr_reader :id, :rooms, :reservations, :hotel_blocks
+
+    def initialize(id:, rooms:, reservations: nil)
+      @id = id
+      @rooms = rooms
+      @reservations ||= reservations
+      @hotel_blocks = []
+    end
+
+    def reserve_room(start_date, end_date, room_id, block_id = nil, discount = nil)
+      #check if room available during those days
+      rooms_available = available_rooms(start_date, end_date)
+      raise ArgumentError if !(rooms_available.include?(room_id)) && block_id == nil
```

Comment:
What if instead of calling `available_rooms` and then checking for inclusion, you called `find_room` and then `is_available?` directly? That would save you calling `is_available?` on all the other rooms.